File: PHP/examples/qr_code_utils_example.php
<?php
/**
 * AllToolkit - QR Code Utilities Example
 *
 * Demonstrates various QR Code generation features and output formats.
 *
 * @package AllToolkit\QrCodeUtils
 */

require_once __DIR__ . '/../qr_code_utils/mod.php';

use AllToolkit\QrCodeGenerator;
use AllToolkit\QrCodeUtils;

echo "========================================\n";
echo "QR Code Utilities Example\n";
echo "========================================\n\n";

// Example 1: Basic ASCII QR Code
echo "Example 1: Basic ASCII QR Code\n";
echo "--------------------------------\n";
echo "Data: 'HELLO WORLD'\n\n";

$qr = new QrCodeGenerator(1, QrCodeGenerator::ERROR_CORRECTION_L);
echo $qr->toAscii("HELLO WORLD");
echo "\n";

// Example 2: Custom ASCII characters
echo "Example 2: Custom ASCII Characters\n";
echo "-----------------------------------\n";
echo "Data: 'TEST' with custom characters\n\n";

echo $qr->toAscii("TEST", "##", "..");
echo "\n";

// Example 3: Numeric data (most compact)
echo "Example 3: Numeric Data (Most Compact)\n";
echo "---------------------------------------\n";
echo "Data: '1234567890'\n\n";

echo $qr->toAscii("1234567890");
echo "\n";

// Example 4: Different versions
echo "Example 4: Different Versions\n";
echo "------------------------------\n";
echo "Data: 'VERSION TEST' with versions 1, 2, 3\n\n";

for ($v = 1; $v <= 3; $v++) {
    $qrV = new QrCodeGenerator($v, QrCodeGenerator::ERROR_CORRECTION_L);
    echo "Version {$v} ({$qrV->getModuleCount()}x{$qrV->getModuleCount()} modules):\n";
    echo $qrV->toAscii("VERSION TEST");
    echo "\n";
}

// Example 5: Different error correction levels
echo "Example 5: Error Correction Levels\n";
echo "-----------------------------------\n";
echo "Data: 'ERROR CORRECTION' with levels L, M, Q, H\n\n";

$levels = [
    QrCodeGenerator::ERROR_CORRECTION_L => 'L (~7%)',
    QrCodeGenerator::ERROR_CORRECTION_M => 'M (~15%)',
    QrCodeGenerator::ERROR_CORRECTION_Q => 'Q (~25%)',
    QrCodeGenerator::ERROR_CORRECTION_H => 'H (~30%)',
];

foreach ($levels as $level => $desc) {
    $qrE = new QrCodeGenerator(2, $level);
    echo "Level {$desc}:\n";
    echo $qrE->toAscii("ERROR CORRECTION");
    echo "\n";
}

// Example 6: Generate SVG output
echo "Example 6: SVG Output\n";
echo "----------------------\n";
echo "Data: 'SVG TEST'\n\n";

$svg = $qr->toSvg("SVG TEST", 4, '#000000', '#ffffff');
echo "SVG generated (length: " . strlen($svg) . " chars)\n";
echo "First 200 characters:\n";
echo substr($svg, 0, 200) . "...\n\n";

// Example 7: Custom colored SVG
echo "Example 7: Custom Colored SVG\n";
echo "------------------------------\n";
echo "Data: 'COLORED' with red on white\n\n";

$svgColored = $qr->toSvg("COLORED", 4, '#ff0000', '#ffffff');
echo "SVG with custom colors generated (length: " . strlen($svgColored) . " chars)\n\n";

// Example 8: Binary matrix output
echo "Example 8: Binary Matrix Output\n";
echo "--------------------------------\n";
echo "Data: 'MATRIX'\n\n";

$matrix = $qr->toMatrix("MATRIX");
echo "Matrix size: " . count($matrix) . "x" . count($matrix[0]) . "\n";
echo "First 5 rows:\n";
for ($i = 0; $i < min(5, count($matrix)); $i++) {
    for ($j = 0; $j < min(10, count($matrix[$i])); $j++) {
        echo $matrix[$i][$j];
    }
    echo "...\n";
}
echo "\n";

// Example 9: Mode detection
echo "Example 9: Mode Detection\n";
echo "--------------------------\n";

$testStrings = [
    "1234567890" => "Numeric",
    "HELLO WORLD" => "Alphanumeric",
    "Hello, World!" => "Byte",
    "test@example.com" => "Byte (email)",
];

foreach ($testStrings as $data => $desc) {
    $mode = QrCodeGenerator::detectMode($data);
    $modeName = QrCodeUtils::getModeName($mode);
    echo "'{$data}' => {$modeName} mode ({$desc})\n";
}
echo "\n";

// Example 10: Utility functions
echo "Example 10: Utility Functions\n";
echo "------------------------------\n";

// Quick ASCII generation
$ascii = QrCodeUtils::toAscii("QUICK", 2, QrCodeGenerator::ERROR_CORRECTION_M);
echo "Quick ASCII generation:\n";
echo $ascii;
echo "\n";

// Quick SVG generation
$svg = QrCodeUtils::toSvg("QUICK SVG", 2, QrCodeGenerator::ERROR_CORRECTION_M, 4, '#0000ff', '#ffffff');
echo "Quick SVG generation (length: " . strlen($svg) . " chars)\n\n";

// Example 11: Save SVG to file
echo "Example 11: Save SVG to File\n";
echo "-----------------------------\n";

$tempFile = sys_get_temp_dir() . '/qrcode_test.svg';
if (QrCodeUtils::saveSvg("SAVE TEST", $tempFile)) {
    echo "SVG saved to: {$tempFile}\n";
    echo "File size: " . filesize($tempFile) . " bytes\n";
    // Clean up
    unlink($tempFile);
} else {
    echo "Failed to save SVG\n";
}
echo "\n";

// Example 12: URL QR Code
echo "Example 12: URL QR Code\n";
echo "------------------------\n";
echo "Data: 'https://github.com/ayukyo/alltoolkit'\n\n";

$qrUrl = new QrCodeGenerator(3, QrCodeGenerator::ERROR_CORRECTION_M);
echo $qrUrl->toAscii("https://github.com/ayukyo/alltoolkit");
echo "\n";

// Example 13: WiFi QR Code (common use case)
echo "Example 13: WiFi QR Code\n";
echo "-------------------------\n";
echo "Data: 'WIFI:T:WPA;S:MyNetwork;P:MyPassword;;'\n\n";

$qrWifi = new QrCodeGenerator(2, QrCodeGenerator::ERROR_CORRECTION_H);
echo $qrWifi->toAscii("WIFI:T:WPA;S:MyNetwork;P:MyPassword;;");
echo "\n";

// Example 14: Email QR Code
echo "Example 14: Email QR Code\n";
echo "--------------------------\n";
echo "Data: 'mailto:test@example.com?subject=Hello'\n\n";

$qrEmail = new QrCodeGenerator(2, QrCodeGenerator::ERROR_CORRECTION_M);
echo $qrEmail->toAscii("mailto:test@example.com?subject=Hello");
echo "\n";

// Example 15: Large data with higher version
echo "Example 15: Large Data (Version 4)\n";
echo "------------------------------------\n";
echo "Data: 'THIS IS A LONGER TEXT THAT REQUIRES A LARGER QR CODE VERSION'\n\n";

$qrLarge = new QrCodeGenerator(4, QrCodeGenerator::ERROR_CORRECTION_L);
echo $qrLarge->toAscii("THIS IS A LONGER TEXT THAT REQUIRES A LARGER QR CODE VERSION");
echo "\n";

echo "========================================\n";
echo "Example Complete!\n";
echo "========================================\n";

File: C#/examples/qr_code_utils_example.cs
using System;
using AllToolkit;

namespace QRCodeUtilsExample
{
    /// <summary>
    /// Example usage of QRCodeUtils
    /// </summary>
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("=== QR Code Utils Examples ===\n");

            Example1_BasicGeneration();
            Example2_TextOutput();
            Example3_SvgOutput();
            Example4_ErrorLevels();
            Example5_ModeDetection();
            Example6_CapacityInfo();
            Example7_CustomCharacters();
            Example8_Validation();
            Example9_VersionInfo();
            Example10_LongerText();
        }

        static void Example1_BasicGeneration()
        {
            Console.WriteLine("Example 1: Basic QR Code Generation");
            Console.WriteLine("-----------------------------------");

            var qr = QRCodeUtils.Generate("HELLO WORLD");
            Console.WriteLine($"Data: {qr.Data}");
            Console.WriteLine($"Version: {qr.Version}");
            Console.WriteLine($"Size: {qr.Width}x{qr.Width} modules");
            Console.WriteLine($"Mode: {qr.Mode}");
            Console.WriteLine($"Error Level: {QRCodeUtils.ErrorLevelToString(qr.ErrorCorrectionLevel)}");
            Console.WriteLine();
        }

        static void Example2_TextOutput()
        {
            Console.WriteLine("Example 2: Text Output");
            Console.WriteLine("----------------------");

            string text = QRCodeUtils.GenerateText("TEST", "██", "  ");
            Console.WriteLine("QR Code (TEST):");
            Console.WriteLine(text);
        }

        static void Example3_SvgOutput()
        {
            Console.WriteLine("Example 3: SVG Output");
            Console.WriteLine("---------------------");

            string svg = QRCodeUtils.GenerateSvg("https://example.com", 4);
            Console.WriteLine("SVG output (first 200 chars):");
            Console.WriteLine(svg.Substring(0, Math.Min(200, svg.Length)) + "...");
            Console.WriteLine();
        }

        static void Example4_ErrorLevels()
        {
            Console.WriteLine("Example 4: Different Error Correction Levels");
            Console.WriteLine("--------------------------------------------");

            string data = "SAME DATA";
            
            foreach (QRErrorCorrectionLevel level in Enum.GetValues(typeof(QRErrorCorrectionLevel)))
            {
                var qr = QRCodeUtils.Generate(data, level);
                Console.WriteLine($"Level {QRCodeUtils.ErrorLevelToString(level)}: Version {qr.Version}");
            }
            Console.WriteLine();
        }

        static void Example5_ModeDetection()
        {
            Console.WriteLine("Example 5: Mode Detection");
            Console.WriteLine("-------------------------");

            string[] testCases = new[] { "12345", "HELLO", "hello", "Test123!", "HELLO WORLD 123" };
            
            foreach (var test in testCases)
            {
                var mode = QRCodeUtils.GetMode(test);
                Console.WriteLine($"'{test}' -> {mode} mode");
            }
            Console.WriteLine();
        }

        static void Example6_CapacityInfo()
        {
            Console.WriteLine("Example 6: Capacity Information");
            Console.WriteLine("-------------------------------");

            string[] testData = new[] { "A", "HELLO", "HELLO WORLD THIS IS A TEST" };
            
            foreach (var data in testData)
            {
                var info = QRCodeUtils.GetCapacityInfo(data);
                Console.WriteLine($"Data: '{data}' ({info.DataLength} chars)");
                Console.WriteLine($"  {info}");
            }
            Console.WriteLine();
        }

        static void Example7_CustomCharacters()
        {
            Console.WriteLine("Example 7: Custom Characters");
            Console.WriteLine("----------------------------");

            // Using different characters for dark/light modules
            string text1 = QRCodeUtils.GenerateText("HI", "##", "  ");
            Console.WriteLine("Using '##' and '  ':");
            Console.WriteLine(text1);

            string text2 = QRCodeUtils.GenerateText("HI", "++", "--");
            Console.WriteLine("Using '++' and '--':");
            Console.WriteLine(text2);
        }

        static void Example8_Validation()
        {
            Console.WriteLine("Example 8: Data Validation");
            Console.WriteLine("--------------------------");

            string[] testCases = new[] { "VALID", "123", "", null };
            
            foreach (var test in testCases)
            {
                bool isValid = QRCodeUtils.ValidateData(test);
                Console.WriteLine($"ValidateData('{test}'): {isValid}");
            }
            Console.WriteLine();
        }

        static void Example9_VersionInfo()
        {
            Console.WriteLine("Example 9: Version Information");
            Console.WriteLine("------------------------------");

            for (int v = 1; v <= 5; v++)
            {
                int size = QRCodeUtils.GetQRCodeSize(v);
                int capNumeric = QRCodeUtils.GetMaxCapacity(v, QRMode.Numeric, QRErrorCorrectionLevel.M);
                int capAlpha = QRCodeUtils.GetMaxCapacity(v, QRMode.Alphanumeric, QRErrorCorrectionLevel.M);
                int capByte = QRCodeUtils.GetMaxCapacity(v, QRMode.Byte, QRErrorCorrectionLevel.M);
                
                Console.WriteLine($"Version {v} ({size}x{size}): Numeric={capNumeric}, Alpha={capAlpha}, Byte={capByte}");
            }
            Console.WriteLine();
        }

        static void Example10_LongerText()
        {
            Console.WriteLine("Example 10: Longer Text");
            Console.WriteLine("-----------------------");

            string url = "https://github.com/ayukyo/alltoolkit";
            var qr = QRCodeUtils.Generate(url, QRErrorCorrectionLevel.M);
            
            Console.WriteLine($"URL: {url}");
            Console.WriteLine($"Length: {url.Length} characters");
            Console.WriteLine($"QR Version: {qr.Version}");
            Console.WriteLine($"QR Size: {qr.Width}x{qr.Width}");
            Console.WriteLine($"Mode: {qr.Mode}");
            Console.WriteLine();

            // Show text representation (smaller)
            string text = QRCodeUtils.GenerateText(url, "█", " ");
            Console.WriteLine("Text representation:");
            Console.WriteLine(text);
        }
    }
}
